src/NuGetGallery.Core/Services/CoreMessageService.cs

```
@@ -192,31 +196,52 @@ protected static void AddOwnersSubscribedToPackagePushedNotification(PackageRegi
             }
         }
 
-        protected void SendMessage(MailMessage mailMessage)
+        protected virtual void SendMessage(MailMessage mailMessage)
         {
-            SendMessage(mailMessage, copySender: false);
+            int attempt = 0;
+            for (;;)
```

Not sure exactly what this means. I assume it means "keep looping infinitely".
I'd rather see `while (true)` than this. Even better, I'd prefer to see this loop rewritten with a clearer condition, e.g. `while (attempt < RetryDelays.Count)`

---

Yes, it's an infinite loop. You're right, `while (true)` is easier to read. I guess some habits die hard.

I'll change the loop to `while (!success)`. The issue with looping until the retry count is exceeded, is that if we want to rethrow the exception and keep the stack trace, we need to keep the exception in a variable, then outside the exception handler do `ExceptionDispatchInfo.Capture(exception).Throw()`. I think it's much cleaner to just `throw;` in the exception handler.